[Woo POS] Add missing reader transitions

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13700
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Added missing reader transitions:
- Fade in / fade out when the modal appears
- Transition between modal states by moving view elements

### Solution

For transitioning between modal states I used the same approach as with in-line message views and applied `matchedGeometryEffect`. It allows synchronizing transitions between same type of elements in different views. 

For modal animations, I worked within `POSModalViewModifier` implementation. I made sure the animation is applied when modal is presented and when `item` changes. 

I also wanted to apply different transitions for modal presentation (push from bottom + opacity) and Reader state changes (element transition + opacity) but couldn't manage to apply a different transition to the content of a modal 🤔 I settled on simpler changes.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Open POS
2. Tap on "Connect your reader"
3. Confirm the modal and the background fades in
4. Confirm transitions are applied between states
5. Confirm the modal and the background fades out

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Changes are user-facing, no unit tests

I manually iterated through all the possible Reader connection alerts to confirm that they work as expected. Attached a video below.

The change also affects other models. Tested:
- Simple product modal
- Exit POS modal
- Capture payment modal

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Simulated iteration through all the reader states

https://github.com/user-attachments/assets/747bfd52-6933-4a8d-a8be-5e0e80cb733b

### Connection flow

https://github.com/user-attachments/assets/813d064a-887f-4f13-b678-828a8f11e7ae

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Animate modal presentation

staskus committed:
Animate modal transitions when item changes

dangermattic commented:
<!--
  0 Errors
  3 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/221">20.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Synchronize title, description, icon and buttons in POS Reader alerts in animated transitions

staskus committed:
Center non-retryable error

staskus committed:
Group progress view title and subtitle together for animation

staskus committed:
Rename messageTransitionId to contentTransitionId

staskus committed:
Only use matchedGeometry for scanningText in multiple reader view

joshheald commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Connection Alerts/PointOfSaleCardPresentPaymentFoundMultipleReadersView.swift:
I don't think we should include the titles and text in the animation transitions for this view... or at least not using position.

It's too different from the other alerts, because the titles have to cross the image as it's fading out, and then go under the list. It's just a bit much for something which should be subtle, a jump-cut might be better.

staskus replied:
Good point. I tried to tweak it in a few different ways but not matching a geometry at all is the best option 👍

staskus committed:
Merge branch 'trunk' into task/13700-pos-m2-add-missing-card-reader-transitions

joshheald committed:
13700 Scale and fade transition for modals

joshheald committed:
13700 Don’t animate multiple reader titles/content

These are too different to really work well as a transition from the other card connection modals.

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSModalViewModifier.swift:
```diff
@@ -17,11 +17,12 @@ struct POSRootModalViewModifier: ViewModifier {
                     .cornerRadius(24)
                     .shadow(color: Color.black.opacity(0.08), radius: 24, x: 0, y: 8)
-                    .transition(.scale)
+                    .transition(.opacity)
                     .zIndex(1)
                     .padding()
             }
         }
+        .animation(.default, value: modalManager.isPresented)
```
Sooo the fade in felt a bit odd to me.

I can see why you opted for something simple, but I think I've cracked it:

https://github.com/user-attachments/assets/aa00f01f-e1dd-4fbd-b823-dcd57eae0b31

#13788 for the details – feel free to either merge or ignore, I see it as part of this PR, it was just easier to communicate with a new branch.

staskus replied:
Thanks! I'll check it out. Definetely looks better 👍

staskus replied:
Looks and works much better 👏 . The changes make sense!

joshheald reviewed (approved):
Looks good. I went down a rabbit hole and worked on the modal appear/disappear transition, see what you think.

One suggestion about not transitioning the multiple readers views.

Thanks for really making it feel a lot better!
